### ISSUE TITLE
Magisk unable to properly sign boot.img after repacking on Android Q

### ISSUE THREAD

**Issue by optimumpr** [topjohnwu/Magisk#2238]
Magisk signing is not recognized on locked bootloader on Oneplus 5/5T. Installation is successful. If image is previously signed, Magisk resigns the image, but on boot, there is red warning: your device is corrupted. On unlocked bootloader the phone boots fine. This happens with every custom Q rom.

**Comment by alexa-v2** [comment]
The signing magisk does is not intended to ever work on a locked bootloader. You have to have an OEM approved key to sign with

The general assumption with magisk is that you have an unlocked bootloader

**Comment by optimumpr** [comment]
Signing worked fine on Pie with locked bootloader with the same key. If kernel is signed with another key, Oneplus 5 boots fine with yellow screen: your device has loaded a different operating system.

**Comment by alexa-v2** [comment]
TBH if that actually works it means they don't implement AVB correctly but that's another story m

Nonetheless this scenario is not designed to work and if it does it's the OEM implementation not magisk problem

**Comment by optimumpr** [comment]
No. I think the reason is that Android Q has a different implementation of AVB, and perhaps, Magisk should follow it. 

By the way, congrats on your Bromite.

**Comment by optimumpr** [comment]
By the way, is there a way to make changes in kernel sources to have it Magisk ready? This way, I can perhaps change Magisk script not to touch kernel.

**Comment by topjohnwu** [comment]
Signing never allows using custom images on locked bootloaders. Please do not do so or you can easily brick your device.

**Comment by osm0sis** [comment]
If we're talking about AVBv1 the answer is likely that -Xnodex2oat is intentionally removed from Android 10 by AOSP. (Though poorly removed since it still shows up in the dalvik --help for now). I plan to push a commit in a PR to fix this soon.

If your device switched to AVBv2 with Android 10 then there's no way to sign these on device since the AOSP AVBv2 signing tool is in Python, and so this can't be supported.

**Comment by optimumpr** [comment]
I make my own roms. So, it is AVBv1. Thank you.

**Comment by osm0sis** [comment]
This should do the trick once merged: https://github.com/topjohnwu/Magisk/pull/2248

**Comment by optimumpr** [comment]
@osm0sis 

Thanks for the commit. It still doesn't work. 

Can you point me to the changes I could make in kernel sources (instead of applying hexes on compiled kernel) to make it Magisk ready? This way, I can modify Magisk script not to touch kernel, but do the changes in data.

**Comment by osm0sis** [comment]
It's not merged yet.

**Comment by optimumpr** [comment]
I took those commits and compiled Magisk myself.
Could it be because Android 10 has /system/system/, as opposed to /system/?

Anyway, is there any way to incorporate Magisk changes on kernel side in sources?

**Comment by osm0sis** [comment]
/system/system (actually called system-as-root or SAR) is already handled by Magisk. You can see in my PR comments that others on Android 10 have AVBv1 working after they too manually add my commit, so perhaps there's some issue with your methodology?

**Comment by optimumpr** [comment]
If not that, then is it possible that perhaps the latest update to the OnePlus 5/T bootloader simply doesn't allow non-OEM signed images anymore?

I have found the reason. I have updated my TWRP, signed it properly, installed and got a red screen - your device is corrupted. This was strange, because an older version of TWRP (also self-signed works fine). Then I changed my PC date to August 2019 (the timestamp of the older TWRP, resigned the new TWRP and was able to boot. Indeed, Oneplus has made changes to its bootloaders and self-signing is not allowed, but only after a certain date. The only question remains: how is it that a boot image compiled and signed with the rom boots?
